### PR TITLE
Issue #5310

### PR DESCRIPTION
**Pull request by cjrosa**
Modified the `MVCRazorParser ` to allow for the use of ModelExpression as the value type for a TagHelper that utilizes an `HtmlAttributeNameAttribute `with a DictionaryAttributePrefix.

As a result, the following can be supported:

`<div prefix-test="@Model" ></div>`

By way of...

```
        [HtmlAttributeName(DictionaryAttributePrefix = "prefix-")]
        public IDictionary<string, ModelExpression> PrefixValues { get; set; } = new Dictionary<string, ModelExpression>();
```
As for test cases... New BASELINES have been generated and expected results have been adjusted where applicable.  The TagHelpersWebsite makes use of a newly added DictionaryAttributePrefix-based TagHelper and all functional tests pass.  The site renders correctly.

### CODE REVIEW

File: src/Microsoft.AspNetCore.Mvc.Razor.Host/MvcRazorParser.cs [hunk 1]
```diff
@@ -87,29 +87,7 @@ public class MvcRazorParser : RazorParser
...
-                            attributeDescriptor.TypeName,
-                            _modelExpressionTypeName,
-                            StringComparison.Ordinal))
-                    {
-                        errorSink.OnError(
-                            SourceLocation.Undefined,
-                            Resources.FormatMvcRazorParser_InvalidPropertyType(
-                                descriptor.TypeName,
-                                attributeDescriptor.Name,
-                                _modelExpressionTypeName),
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Razor.Host/MvcRazorParser.cs]: Remove the `_modelExpressionTypeName` field from this class. Leave the constructor parameter and its `null` check alone but don't use it.

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/DictionaryPrefixTestTagHelper.cs [hunk 2]
```diff
@@ -0,0 +1,17 @@
+﻿using Microsoft.AspNetCore.Mvc.ViewFeatures;
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/DictionaryPrefixTestTagHelper.cs]: Add :copyright: text

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/DictionaryPrefixTestTagHelper.cs [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+using Microsoft.AspNetCore.Razor.TagHelpers;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Host.Test
+{
+    [HtmlTargetElement(Attributes = "prefix-*")]
+    public class DictionaryPrefixTestTagHelper : TagHelper
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/DictionaryPrefixTestTagHelper.cs]: The generated code below shows this tag helper is not bound. Please figure out why and fix it.

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/Internal/ViewComponentTagHelperChunkVisitorTest.cs [hunk 4]
```diff
@@ -7,6 +7,7 @@
 using Microsoft.AspNetCore.Razor.Chunks;
 using Microsoft.AspNetCore.Razor.CodeGenerators;
 using Xunit;
+using System;
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/Internal/ViewComponentTagHelperChunkVisitorTest.cs]: Please revert.

File: test/WebSites/TagHelpersWebSite/Startup.cs [hunk 5]
```diff
@@ -20,6 +20,8 @@ public void Configure(IApplicationBuilder app)
         {
             app.UseCultureReplacer();
 
+            app.UseDeveloperExceptionPage();
+
```
Comment (dougbu) [inline test/WebSites/TagHelpersWebSite/Startup.cs]: Please revert.

File: test/WebSites/TagHelpersWebSite/TagHelpers/DictionaryPrefixTestTagHelper.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
+﻿using Microsoft.AspNetCore.Mvc.ViewFeatures;
```
Comment (dougbu) [inline test/WebSites/TagHelpersWebSite/TagHelpers/DictionaryPrefixTestTagHelper.cs]: Add :copyright: text

File: test/WebSites/TagHelpersWebSite/project.json [hunk 7]
```diff
@@ -19,7 +20,7 @@
   "frameworks": {
     "net451": {},
     "netcoreapp1.0": {
-      "dependencies": {
+      "dependencies": {
```
Comment (dougbu) [inline test/WebSites/TagHelpersWebSite/project.json]: Please revert all changes to this file.

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/TestFiles/Input/ModelExpressionTagHelper.cshtml [hunk 8]
```diff
@@ -1,6 +1,9 @@
...
 
 @addTagHelper Microsoft.AspNetCore.Mvc.Razor.InputTestTagHelper, Microsoft.AspNetCore.Mvc.Razor.Host.Test
+@addTagHelper Microsoft.AspNetCore.Mvc.Razor.DictionaryPrefixTestTagHelper, Microsoft.AspNetCore.Mvc.Razor.Host.Test
 
 <input-test for="Now" />
-<input-test for="@Model" />
\ No newline at end of file
+<input-test for="@Model" />
+
+<div prefix-test="@Model" ></div>
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/TestFiles/Input/ModelExpressionTagHelper.cshtml]: Please add another element using a `DateTime` property and adding more than one entry to the dictionary e.g. `<span prefix-hour="Hour" prefix-minute="Minute">Text</span>`.

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/Internal/ViewComponentTagHelperChunkVisitorTest.cs (hunk 4 above)
Comment (cjrosa) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/Internal/ViewComponentTagHelperChunkVisitorTest.cs]: This was necessary to allow compilation when GENERATE_BASELINES is [pre-processor] defined.  I removed and changed inline as necessary.

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/Internal/ViewComponentTagHelperChunkVisitorTest.cs [hunk 9]
```diff
@@ -46,7 +46,7 @@ public void Accept_CorrectlyGeneratesCode(IList<Chunk> chunks)
             var resultOutput = writer.GenerateCode();
 
 #if GENERATE_BASELINES
-            if (!string.Equals(expectedOutput, resultOutput, StringComparison.Ordinal))
+            if (!string.Equals(expectedOutput, resultOutput, System.StringComparison.Ordinal))
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/Internal/ViewComponentTagHelperChunkVisitorTest.cs]: Well :ok:. It's a bit ugly but probably better than
``` c#
 #if GENERATE_BASELINES
using System;
#endif
using System.Collections.Generic;
```

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/MvcRazorHostTest.cs [hunk 10]
```diff
@@ -143,31 +143,48 @@ public void MvcRazorHost_GeneratesTagHelperModelExpressionCode_DesignTime()
...
+                    documentAbsoluteIndex: 322,
+                    documentLineIndex: 8,
+                    documentCharacterIndex: 19,
+                    generatedAbsoluteIndex: 3169,
+                    generatedLineIndex: 62,
+                    generatedCharacterIndex: 19,
+                    contentLength: 5),
             };
 
+
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/MvcRazorHostTest.cs]: nit: Please remove the extra newline.

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/MvcRazorParserTest.cs [hunk 11]
```diff
@@ -238,7 +238,7 @@ private class TestableMvcRazorParser : MvcRazorParser
                 RazorParser parser,
                 IReadOnlyList<ChunkTree> chunkTrees,
                 IReadOnlyList<Chunk> defaultInheritedChunks)
-                : base(parser, chunkTrees, defaultInheritedChunks, typeof(ModelExpression).FullName)
+                : base(parser, chunkTrees, defaultInheritedChunks)
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/MvcRazorParserTest.cs]: Revert changes to this file.

File: src/Microsoft.AspNetCore.Mvc.Razor.Host/MvcRazorParser.cs (hunk 1 above)
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Razor.Host/MvcRazorParser.cs]: Missed the second part here.

This is a `public` type and we cannot make breaking changes at this point. Also not worth adding another constructor overload without the unnecessary `modelExpressionTypeName` parameter. So, leave the constructor parameter and the `null` check alone. Just remove the `_modelExpressionTypeName` field -- as you've done.

File: src/Microsoft.AspNetCore.Mvc.Razor.Host/MvcRazorHost.cs [hunk 12]
```diff
@@ -294,7 +294,7 @@ public override RazorParser DecorateRazorParser(RazorParser razorParser, string
             }
 
             var inheritedChunkTrees = GetInheritedChunkTrees(sourceFileName);
-            return new MvcRazorParser(razorParser, inheritedChunkTrees, DefaultInheritedChunks, ModelExpressionType);
+            return new MvcRazorParser(razorParser, inheritedChunkTrees, DefaultInheritedChunks);
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Razor.Host/MvcRazorHost.cs]: Revert changes to this file.

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/DictionaryPrefixTestTagHelper.cs (hunk 3 above)
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/DictionaryPrefixTestTagHelper.cs]: The ModelExpressionTagHelper.cs baseline files are still incorrect. This is very important because the other tag helper used in that test is successfully associated with the target elements.

Suspect the problem is the namespace. It should be `Microsoft.AspNetCore.Mvc.Razor` (no `.Host.Test`) to match what's requested in the `.cshtml` file.

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/DictionaryPrefixTestTagHelper.cs [hunk 13]
```diff
@@ -0,0 +1,20 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Mvc.ViewFeatures;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+using System.Collections.Generic;
```
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/DictionaryPrefixTestTagHelper.cs]: Sort `using`s with `System` ones first. That's about the only non-default Visual Studio setting we've standardized on.

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/TestFiles/Input/ModelExpressionTagHelper.cshtml (hunk 8 above)
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/TestFiles/Input/ModelExpressionTagHelper.cshtml]: thx

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/Internal/ViewComponentTagHelperChunkVisitorTest.cs (hunk 9 above)
Comment (cjrosa) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/Internal/ViewComponentTagHelperChunkVisitorTest.cs]: OK...  Will leave as is

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/MvcRazorHostTest.cs (hunk 10 above)
Comment (cjrosa) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/MvcRazorHostTest.cs]: Done

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/MvcRazorParserTest.cs (hunk 11 above)
Comment (cjrosa) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/MvcRazorParserTest.cs]: As it relates to the MvcRazorParser ~ctor...  Why not make the make the modelExpressionTypeName parameter have a default value = null.  This will allow the other changes to remain intact (i.e. not specifying typeof(ModelExpression).FullName) and will not brake any derivatives of MvcRazorParser.  Thoughts?

File: src/Microsoft.AspNetCore.Mvc.Razor.Host/MvcRazorHost.cs (hunk 12 above)
Comment (cjrosa) [inline src/Microsoft.AspNetCore.Mvc.Razor.Host/MvcRazorHost.cs]: Depends on approach to the ~ctor

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/DictionaryPrefixTestTagHelper.cs (hunk 13 above)
Comment (cjrosa) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/DictionaryPrefixTestTagHelper.cs]: Done

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/DictionaryPrefixTestTagHelper.cs (hunk 3 above)
Comment (cjrosa) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/DictionaryPrefixTestTagHelper.cs]: My apologies, I noticed that the generated file looks different but I'm not sure how to impact the baseline files nor do I know what it's supposed to look like for Dictionary-based TagHelpers.  They are generated utilizing the provided `MvcRazorHost`.  The `TagHelpersWebsite `is functional with the new TagHelper.

File: src/Microsoft.AspNetCore.Mvc.Razor.Host/MvcRazorParser.cs (hunk 1 above)
Comment (cjrosa) [inline src/Microsoft.AspNetCore.Mvc.Razor.Host/MvcRazorParser.cs]: I added the comment below...  gist is using default parameter value

Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Razor.Host/MvcRazorParser.cs]: We avoid optional parameters in almost every case except `private` helpers. See [https://github.com/aspnet/Home/wiki/Engineering-guidelines#async-method-patterns]() for another corner case exception. And, a change is not mandated here -- just nice to have. Leave the signature of the `MvcRazorHost` constructor as it was.

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/Internal/ViewComponentTagHelperChunkVisitorTest.cs (hunk 4 above)
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/Internal/ViewComponentTagHelperChunkVisitorTest.cs]: Got it, thanx.

File: src/Microsoft.AspNetCore.Mvc.Razor.Host/MvcRazorParser.cs (hunk 1 above)
Comment (cjrosa) [inline src/Microsoft.AspNetCore.Mvc.Razor.Host/MvcRazorParser.cs]: Got it...  Done

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/DictionaryPrefixTestTagHelper.cs (hunk 3 above)
Comment (cjrosa) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/DictionaryPrefixTestTagHelper.cs]: Apparently I just need to step away from this for a moment (and get a good night's sleep :)).  All set!!!

File: test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/DictionaryPrefixTestTagHelper.cs (hunk 13 above)
Comment (dougbu) [inline test/Microsoft.AspNetCore.Mvc.Razor.Host.Test/DictionaryPrefixTestTagHelper.cs]: Not done and the order is also off in the other `DictionaryPrefixTestTagHelper` class.

File: test/WebSites/TagHelpersWebSite/TagHelpers/DictionaryPrefixTestTagHelper.cs [hunk 14]
```diff
@@ -0,0 +1,33 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using Microsoft.AspNetCore.Mvc.Rendering;
+using Microsoft.AspNetCore.Mvc.ViewFeatures;
+using Microsoft.AspNetCore.Razor.TagHelpers;
+using System.Collections.Generic;
+
+namespace Microsoft.AspNetCore.Mvc.Razor.Host.Test
```
Comment (dougbu) [inline test/WebSites/TagHelpersWebSite/TagHelpers/DictionaryPrefixTestTagHelper.cs]: `namespace TagHelpersWebSite.TagHelpers` to match others in this directory.

File: src/Microsoft.AspNetCore.Mvc.Razor.Host/MvcRazorParser.cs [hunk 15]
```diff
@@ -31,7 +30,7 @@ public class MvcRazorParser : RazorParser
         /// from parsed pages from _ViewImports files.</param>
         /// <param name="defaultInheritedChunks">The <see cref="IReadOnlyList{Chunk}"/> inherited by
         /// default by all Razor pages in the application.</param>
-        /// <param name="modelExpressionTypeName">The full name of the model expression <see cref="Type"/>.</param>
+        /// <param name="modelExpressionTypeName">The full name of the model expression <see cref="Type"/>.</param>
```
Comment (dougbu) [inline src/Microsoft.AspNetCore.Mvc.Razor.Host/MvcRazorParser.cs]: Not mandatory but remove trailing whitespace added here if you open the file again.

**Comment by dougbu** [comment]
@cjrosa this is ready to go except for the `using` ordering in new classes. Put `System` namespaces first.

**Comment by cjrosa** [comment]
@dougbu Thanks for your patience and guidance!  I'd like to start contributing more as time permits.

**Comment by dougbu** [comment]
60d600d
